zokrates_core/src/static_analysis/reducer/constants_writer.rs

```diff
@@ -0,0 +1,163 @@
+    }
+
+    fn update_program(&mut self) {
+        let mut p = TypedProgram {
+            main: "".into(),
+            modules: BTreeMap::default(),
+        };
+        std::mem::swap(&mut self.program, &mut p);
+        self.program = ConstantsReader::with_constants(&self.constants).read_into_program(p);
+    }
+
+    fn update_symbol_declaration(
+        &self,
+        d: TypedSymbolDeclaration<'ast, T>,
+    ) -> TypedSymbolDeclaration<'ast, T> {
+        ConstantsReader::with_constants(&self.constants).read_into_symbol_declaration(d)
+    }
+}
+
+impl<'ast, T: Field> ResultFolder<'ast, T> for ConstantsWriter<'ast, T> {
+    type Error = Error;
+
+    fn fold_module_id(
+        &mut self,
+        id: OwnedTypedModuleId,
+    ) -> Result<OwnedTypedModuleId, Self::Error> {
+        // anytime we encounter a module id, visit the corresponding module if it hasn't been done yet
+        if !self.treated(&id) {
+            let current_m_id = self.change_location(id.clone());
+            // I did not find a way to achieve this without cloning the module. Assuming we do not clone:
+            // to fold the module, we need to consume it, so it gets removed from the modules
+            // but to inline the calls while folding the module, all modules must be present
+            // therefore we clone...
+            // this does not lead to a module being folded more than once, as the first time
+            // we change location to this module, it's added to the `treated` set
+            let m = self.program.modules.get(&id).cloned().unwrap();
+            let m = self.fold_module(m)?;
+            self.program.modules.insert(id.clone(), m);
+            self.change_location(current_m_id);
+        }
+        Ok(id)
+    }
+
+    fn fold_symbol_declaration(
+        &mut self,
+        s: TypedSymbolDeclaration<'ast, T>,
+    ) -> Result<TypedSymbolDeclaration<'ast, T>, Self::Error> {
+        // before we treat the symbol, propagate the constants into it, as may be using constants defined earlier in this module.
+        let s = self.update_symbol_declaration(s);
+
+        let s = fold_symbol_declaration(self, s)?;
+
+        // after we treat the symbol, propagate again, as treating this symbol may have triggered checking another module, resolving new constants which this symbol may be using.
+        Ok(self.update_symbol_declaration(s))
+    }
+
+    fn fold_constant_symbol_declaration(
+        &mut self,
+        d: TypedConstantSymbolDeclaration<'ast, T>,
+    ) -> Result<TypedConstantSymbolDeclaration<'ast, T>, Self::Error> {
+        let id = self.fold_canonical_constant_identifier(d.id)?;
+
+        match d.symbol {
+            TypedConstantSymbol::Here(c) => {
+                let c = self.fold_constant(c)?;
+
+                use crate::typed_absy::{DeclarationSignature, TypedFunction, TypedStatement};
+
+                // wrap this expression in a function
+                let wrapper = TypedFunction {
```

Comment:
Ah I see, this is the hacky part. I think at some point we should have some kind of expression evaluator so we don't have to wrap expressions in a function to reduce. We can roll with this until then I guess.

Reply:
will open an issue for this

Reply:
#1021